File: leetcode/algorithms/decode_ways/decode_ways.cc
class Solution {
public:
    int numDecodings(string s) {
        if (s.empty() || s == "0" || s[0] == '0') {
            return 0;
        }

        int n = s.size();
        vector<int> ways(n, 0);

        ways[0] = 1;
        ways[1] = s[1] != '0' ? 1 : 0;

        if (isValid(s.substr(0, 2))) {
            ++ways[1];
        }

        for (int i = 2; i < n; ++i) {
            if (s[i] != '0') {
                ways[i] = ways[i - 1];
            }

            if (isValid(s.substr(i - 1, 2))) {
                ways[i] += ways[i - 2];
            }
        }

        return ways[n - 1];
    }

    bool isValid(string s) {
        int number = stoi(s);
        return number >= 10 && number <= 26;
    }
};

File: leetcode/algorithms/decode_ways/decode_ways.py
class Solution(object):
    def numDecodings(self, s):
        n = len(s)

        if n == 0 or s == '0':
            return 0

        self.memory = [-1] * n
        return self.ways(s, 0)

    def ways(self, s, start):
        n = len(s)

        if start == n:
            return 1

        if s[start] == '0':
            return 0

        if self.memory[start] != -1:
            return self.memory[start]

        solution = self.ways(s, start + 1)

        if start < n - 1 and  10 <= int(s[start:start+2]) <= 26:
            solution += self.ways(s, start + 2)

        self.memory[start] = solution
        return self.memory[start]
